File: python/975_OddEvenJump.py
from unittest import TestCase
# https://leetcode.com/problems/odd-even-jump


class OddEvenJump(TestCase):
    def oddEvenJumps(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        n, res = len(A), 1
        next_lower, next_higher, stack = [0]*n, [0]*n, []       # 0 is impossible
        for v, i in sorted((v, i) for i, v in enumerate(A)):
            while stack and stack[-1] < i:
                next_higher[stack.pop()] = i
            stack.append(i)

        stack = []
        for v, i in sorted((-v, i) for i, v in enumerate(A)):
            while stack and stack[-1] < i:
                next_lower[stack.pop()] = i
            stack.append(i)

        good = [(True, True)]*n
        for i in range(n-2, -1, -1):
            odd = next_higher[i] and good[next_higher[i]][1]
            even = next_lower[i] and good[next_lower[i]][0]
            if odd: res += 1
            good[i] = odd, even
        return res

    def test1(self):
        self.assertEqual(2, self.oddEvenJumps([10,13,12,14,15]))

    def test2(self):
        self.assertEqual(3, self.oddEvenJumps([2,3,1,1,4]))

    def test3(self):
        self.assertEqual(3, self.oddEvenJumps([5,1,3,4,2]))

File: kotlin/src/main/kotlin/975.OddEvenJump.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

// https://leetcode.com/problems/odd-even-jump


class OddEvenJump {
  fun oddEvenJumps(A: IntArray): Int {
    var (res, n) = 1 to A.size
    val order = TreeSet<Pair<Int, Int>>(compareBy({it.second}, {it.first}))
    val good = hashMapOf<Int, Pair<Boolean, Boolean>>()
    order.add(n-1 to A[n-1])
    good[n-1] = true to true
    for(i in n-2 downTo 0){
      val v = A[i]
      val r = order.ceiling(i to v)
      val odd = r != null && good[r.first]!!.second
      var l = order.floor(n to v)
      if(l != null)
        l = order.ceiling(0 to l.second)
      val even = l != null && good[l.first]!!.first
      if(odd) res ++
      good[i] = odd to even
      order.add(i to v)
    }
    return res
  }

  @Test
  fun test1(){
    assertEquals(2, oddEvenJumps(intArrayOf(10,13,12,14,15)))
  }

  @Test
  fun test2(){
    assertEquals(3, oddEvenJumps(intArrayOf(2,3,1,1,4)))
  }

  @Test
  fun test3(){
    assertEquals(3, oddEvenJumps(intArrayOf(5,1,3,4,2)))
  }
}

